### PR TITLE
Features/token ablations

### PR DESCRIPTION
**Pull request by curt-tigges**
# Description

I've added separate functions for getting loss and logit diff. Can ignore the notebook changes (just some random tests) and the period ablation stuff (which is in progress).

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have not rewritten tests relating to key interfaces which would affect backward compatibility
- [ ] I have followed the style guide below

# Style guide
- My code follows the [Black](https://pypi.org/project/black/) format
- I have minimised type warnings from [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance)
- Cull any redundant and unused code
- Cache results of slow operations in as flexible of a format as possible, in a flat structure inside results/cache and the file name contains all necessary identifying information (such as model name and dataset used) to avoid accidental overwriting.
- Use intuitive abstractions, making clear the public entrypoints to a given class
- Function and variable names should be long enough to be descriptive

### CODE REVIEW

**Comment by curt-tigges** [comment]
@ojh31 I noticed that somewhere in the process of rebasing, the function mask_loss_at_next_positions disappeared. What was the intended replacement for this?

**Comment by ojh31** [comment]
> @ojh31 I noticed that somewhere in the process of rebasing, the function mask_loss_at_next_positions disappeared. What was the intended replacement for this?

It was replaced by [this](https://github.com/curt-tigges/exploring-summarization/blob/8d397f9f89a188fdee7e89d15dd99cfbdb12dac3/utils/datasets.py#L144). The idea is we should really just cache all of the losses and then after loading the cache you can mask however you like. I tried to unite the different cases we want to mask into one function.

File: utils/tokenwise_ablation.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ def compute_ablation_modified_metric(
     all_positions: bool = False,
     metric: Literal["logits", "loss"] = "logits",
     device: torch.device = DEFAULT_DEVICE,
-    overwrite: bool = False,
+    cached: bool = True,
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:200]: Should we not remove this function now?

File: utils/tokenwise_ablation.py [hunk 2]
```diff
@@ -422,6 +422,269 @@ def compute_ablation_modified_metric(
...
+    if cached and file.exists():
+        return torch.load(file.path)
+    if layers_to_ablate == "all":
+        layers_to_ablate = list(range(model.cfg.n_layers))
+    experiment_names = [
+        "orig",
+        "ablated",
+    ]
+    # get length of prompts in dataset
+    item_len = len(data_loader.dataset[0]["tokens"])
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:484]: Good catch, this is safer than using `n_ctx`

File: utils/tokenwise_ablation.py [hunk 3]
```diff
@@ -422,6 +422,269 @@ def compute_ablation_modified_metric(
...
+        # get the loss for each token when run with hooks
+        hooked_loss = model(
+            batch_tokens, return_type="loss", prepend_bos=False, loss_per_token=True
+        )
+        # concatenate column of 0s
+        hooked_loss = torch.cat(
+            [torch.zeros((hooked_loss.shape[0], 1)).to(device), hooked_loss], dim=1
+        )
+        loss_diff = hooked_loss - orig_metric
+        ablated_metric = mask_loss_at_next_positions(
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:544]: Should remove this reference to the old masking function and cache all of the loss diffs here

File: utils/tokenwise_ablation.py [hunk 4]
```diff
@@ -422,6 +422,269 @@ def compute_ablation_modified_metric(
...
+            batch_idx * batch_size : (batch_idx + 1) * batch_size,
+        ] = ablated_metric.cpu()
+
+        model.reset_hooks()
+
+    torch.save(output, file.path)
+    return output
+
+
+def compute_ablation_modified_logit_diff(
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:561]: Ideally it would be nice to have some sort of documentation where we compare and contrast the similarly named functions in this file

**Review by ojh31 — CHANGES_REQUESTED** [review]
Just need to remove the reference to `mask_loss_at_next_positions` and then this is good to merge
